### DISCUSSION TITLE
How to completely exit the engine execution?

### DISCUSSION THREAD

**Discussion by eduard-nomi** [hyperjumptech/grule-rule-engine#330] [Q&A] [answered]
Is there a way to stop execution from the rule?
Not Retract + Complete but immediately exit without further execution when a rule matched.
It's fine that `when` parts executed, I just don't need evaluation of next rules.
Is it possible to do?

```
rule Stop "stop"  salience 10 {
when <something>
then 
     Result = "completed with errors";
     Stop();
}

rule Next "next, not executed" {
...
}
```

**Comment by newm4n** [comment, marked as answer]
Function Complete() will stop the engine execution loop and thus stops grule.

https://github.com/hyperjumptech/grule-rule-engine/blob/master/docs/en/Function_en.md#complete

On this line ( https://github.com/hyperjumptech/grule-rule-engine/blob/6cbdf38043bb1d70750b0a72ebcc2567bafb65eb/engine/GruleEngine.go#L208 ) you can see that a condition check on the data context IsComplete() method will break the engine loop and thus finishing the engine execution.

**Reply by newm4n** [reply]
As you can see in this line, This built-in function calls  the  datacontext's Complete() function, which latter  checked by the engine loop.
https://github.com/hyperjumptech/grule-rule-engine/blob/6cbdf38043bb1d70750b0a72ebcc2567bafb65eb/ast/BuiltInFunctions.go#L46

So, 

```
then
    Complete();
```

will stop the engine execution.

**Reply by eduard-nomi** [reply]
Thanks for quick answer. As I understand from the doc Complete will stop current loop only and start again evaluating all rules from the start?
Complete will cause the engine to stop processing further rules **in its current cycle**.
If it's correct my question is how to stop the engine completely and prevent further processing.

**Reply by newm4n** [reply]
Looking back to the source code, the Complete() function will indeed stop the engine execution. There won't be any further evaluation loop. It stop and exit the Execute function.

**Reply by eduard-nomi** [reply]
Thank you very much! 👍
